File: backend/app/api/plans.py
"""Authenticated Stage 2 Plan and Task query endpoints."""

from datetime import date
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Header, Query, status

from app.api.dependencies import (
    get_current_user,
    get_plan_query_service,
    get_task_adjustment_service,
)
from app.core.security import AuthenticatedUser
from app.schemas.enums import PlanStatus, TaskStatus
from app.schemas.errors import ErrorResponse
from app.schemas.plans import (
    ActivePlanResponse,
    PlanListResponse,
    PlanSourcesResponse,
    TaskAdjustmentCreateRequest,
    TaskAdjustmentDecisionRequest,
    TaskAdjustmentProposalResponse,
    TaskChecklistUpdateRequest,
    TaskDetailResponse,
    TaskEditRequest,
    TaskEditResponse,
    TaskListResponse,
    TaskUpdateRequest,
    TaskUpdateResponse,
    TaskVerificationRequest,
)
from app.services.plans import PlanQueryService
from app.services.task_adjustments import TaskAdjustmentService

plans_router = APIRouter(prefix="/plans", tags=["plans"])
tasks_router = APIRouter(prefix="/tasks", tags=["tasks"])
task_adjustments_router = APIRouter(
    prefix="/task-adjustment-proposals", tags=["task-adjustments"]
)


@plans_router.get(
    "/active",
    response_model=ActivePlanResponse,
    responses={401: {"model": ErrorResponse}, 404: {"model": ErrorResponse}},
)
async def get_active_plan(
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
) -> ActivePlanResponse:
    return await service.get_active(current_user.id)


@plans_router.get(
    "",
    response_model=PlanListResponse,
    responses={401: {"model": ErrorResponse}},
)
async def list_plans(
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
    status: PlanStatus | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    cursor: UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
) -> PlanListResponse:
    return await service.list_plans(
        user_id=current_user.id,
        status=status.value if status else None,
        date_from=date_from,
        date_to=date_to,
        cursor=cursor,
        limit=limit,
    )


@plans_router.get(
    "/{plan_id}/sources",
    response_model=PlanSourcesResponse,
    responses={401: {"model": ErrorResponse}, 404: {"model": ErrorResponse}},
)
async def get_plan_sources(
    plan_id: UUID,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
) -> PlanSourcesResponse:
    return await service.get_sources(plan_id, current_user.id)


@plans_router.get(
    "/{plan_id}",
    response_model=ActivePlanResponse,
    responses={401: {"model": ErrorResponse}, 404: {"model": ErrorResponse}},
)
async def get_plan(
    plan_id: UUID,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
) -> ActivePlanResponse:
    return await service.get_plan(plan_id, current_user.id)


@tasks_router.get(
    "",
    response_model=TaskListResponse,
    responses={401: {"model": ErrorResponse}},
)
async def list_tasks(
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
    scheduled_date: Annotated[date | None, Query(alias="date")] = None,
    state: TaskStatus | None = None,
    plan_id: UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 50,
) -> TaskListResponse:
    return await service.list_tasks(
        user_id=current_user.id,
        scheduled_date=scheduled_date,
        state=state.value if state else None,
        plan_id=plan_id,
        limit=limit,
    )


@tasks_router.patch(
    "/{task_id}",
    response_model=TaskUpdateResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
        422: {"model": ErrorResponse},
    },
)
async def update_task(
    task_id: UUID,
    payload: TaskUpdateRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
) -> TaskUpdateResponse:
    return await service.update_task(
        task_id=task_id,
        user_id=current_user.id,
        payload=payload,
    )


@tasks_router.patch(
    "/{task_id}/checklist",
    response_model=TaskUpdateResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
        422: {"model": ErrorResponse},
    },
)
async def update_task_checklist(
    task_id: UUID,
    payload: TaskChecklistUpdateRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
) -> TaskUpdateResponse:
    return await service.update_task_checklist(
        task_id=task_id,
        user_id=current_user.id,
        payload=payload,
    )


@tasks_router.patch(
    "/{task_id}/verification",
    response_model=TaskUpdateResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
        422: {"model": ErrorResponse},
    },
)
async def verify_task(
    task_id: UUID,
    payload: TaskVerificationRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[PlanQueryService, Depends(get_plan_query_service)],
) -> TaskUpdateResponse:
    return await service.verify_task(
        task_id=task_id,
        user_id=current_user.id,
        payload=payload,
    )


@tasks_router.get(
    "/{task_id}",
    response_model=TaskDetailResponse,
    responses={401: {"model": ErrorResponse}, 404: {"model": ErrorResponse}},
)
async def get_task_detail(
    task_id: UUID,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[TaskAdjustmentService, Depends(get_task_adjustment_service)],
) -> TaskDetailResponse:
    return await service.get_detail(task_id=task_id, user_id=current_user.id)


@tasks_router.patch(
    "/{task_id}/details",
    response_model=TaskEditResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
        422: {"model": ErrorResponse},
    },
)
async def edit_task_details(
    task_id: UUID,
    payload: TaskEditRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[TaskAdjustmentService, Depends(get_task_adjustment_service)],
    idempotency_key: Annotated[
        str, Header(alias="Idempotency-Key", min_length=1, max_length=64)
    ],
) -> TaskEditResponse:
    return await service.edit_manual(
        task_id=task_id,
        user_id=current_user.id,
        payload=payload,
        idempotency_key=idempotency_key,
    )


@tasks_router.post(
    "/{task_id}/adjustment-proposals",
    status_code=status.HTTP_201_CREATED,
    response_model=TaskAdjustmentProposalResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
        422: {"model": ErrorResponse},
        502: {"model": ErrorResponse},
    },
)
async def create_task_adjustment_proposal(
    task_id: UUID,
    payload: TaskAdjustmentCreateRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[TaskAdjustmentService, Depends(get_task_adjustment_service)],
    idempotency_key: Annotated[
        str, Header(alias="Idempotency-Key", min_length=1, max_length=64)
    ],
) -> TaskAdjustmentProposalResponse:
    return await service.propose(
        task_id=task_id,
        user_id=current_user.id,
        payload=payload,
        idempotency_key=idempotency_key,
    )


@task_adjustments_router.post(
    "/{adjustment_id}/confirm",
    response_model=TaskEditResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
        422: {"model": ErrorResponse},
    },
)
async def confirm_task_adjustment(
    adjustment_id: UUID,
    payload: TaskAdjustmentDecisionRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[TaskAdjustmentService, Depends(get_task_adjustment_service)],
) -> TaskEditResponse:
    return await service.confirm(
        adjustment_id=adjustment_id,
        user_id=current_user.id,
        version=payload.version,
    )


@task_adjustments_router.post(
    "/{adjustment_id}/reject",
    response_model=TaskAdjustmentProposalResponse,
    responses={
        401: {"model": ErrorResponse},
        404: {"model": ErrorResponse},
        409: {"model": ErrorResponse},
    },
)
async def reject_task_adjustment(
    adjustment_id: UUID,
    payload: TaskAdjustmentDecisionRequest,
    current_user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    service: Annotated[TaskAdjustmentService, Depends(get_task_adjustment_service)],
) -> TaskAdjustmentProposalResponse:
    return await service.reject(
        adjustment_id=adjustment_id,
        user_id=current_user.id,
        version=payload.version,
    )

File: frontend/src/api/plans.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiRequest } from "./client";
import type {
  ActivePlanResponse,
  MeResponse,
  TaskAdjustmentProposalResponse,
  TaskChecklistUpdateRequest,
  TaskDetailResponse,
  TaskEditRequest,
  TaskEditResponse,
  TaskUpdateRequest,
  TaskUpdateResponse,
  TaskVerificationRequest,
} from "./types";

export interface PlanListResponse {
  items: ActivePlanResponse[];
  next_cursor: string | null;
}

export function useActivePlan() {
  return useQuery({
    queryKey: ["plans", "active"],
    queryFn: () => apiRequest<ActivePlanResponse>("/api/v1/plans/active"),
    retry: false,
  });
}

export function usePlan(planId: string | undefined) {
  return useQuery({
    queryKey: ["plans", planId],
    queryFn: () => apiRequest<ActivePlanResponse>(`/api/v1/plans/${planId}`),
    enabled: planId !== undefined,
    retry: false,
  });
}

export function usePlans() {
  return useQuery({
    queryKey: ["plans"],
    queryFn: () => apiRequest<PlanListResponse>("/api/v1/plans"),
    retry: false,
  });
}

export function useUpdateTask() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ taskId, payload }: { taskId: string; payload: TaskUpdateRequest }) =>
      apiRequest<TaskUpdateResponse>(`/api/v1/tasks/${taskId}`, {
        method: "PATCH",
        body: payload,
      }),
    onSuccess: (result) => {
      updateMeTaskCache(qc, result.task);
      qc.invalidateQueries({ queryKey: ["plans"] });
      qc.invalidateQueries({ queryKey: ["tasks"] });
      qc.invalidateQueries({ queryKey: ["me"] });
    },
  });
}

export function useUpdateTaskChecklist() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      taskId,
      payload,
    }: {
      taskId: string;
      payload: TaskChecklistUpdateRequest;
    }) => apiRequest<TaskUpdateResponse>(`/api/v1/tasks/${taskId}/checklist`, {
      method: "PATCH",
      body: payload,
    }),
    onSuccess: (result) => {
      updateMeTaskCache(qc, result.task);
      invalidateTaskData(qc, result.task.task_id);
    },
  });
}

export function useVerifyTask() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      taskId,
      payload,
    }: {
      taskId: string;
      payload: TaskVerificationRequest;
    }) => apiRequest<TaskUpdateResponse>(`/api/v1/tasks/${taskId}/verification`, {
      method: "PATCH",
      body: payload,
    }),
    onSuccess: (result) => {
      updateMeTaskCache(qc, result.task);
      invalidateTaskData(qc, result.task.task_id);
    },
  });
}

export function useTaskDetail(taskId: string | undefined) {
  return useQuery({
    queryKey: ["tasks", taskId, "detail"],
    queryFn: () => apiRequest<TaskDetailResponse>(`/api/v1/tasks/${taskId}`),
    enabled: taskId !== undefined,
    retry: false,
  });
}

function invalidateTaskData(qc: ReturnType<typeof useQueryClient>, taskId: string): void {
  qc.invalidateQueries({ queryKey: ["tasks", taskId, "detail"] });
  qc.invalidateQueries({ queryKey: ["plans"] });
  qc.invalidateQueries({ queryKey: ["me"] });
}

function updateMeTaskCache(
  qc: ReturnType<typeof useQueryClient>,
  task: TaskUpdateResponse["task"],
): void {
  qc.setQueryData<MeResponse>(["me"], (current) => {
    if (current === undefined) return current;
    return {
      ...current,
      today_tasks: current.today_tasks.map((item) =>
        item.task_id === task.task_id ? task : item
      ),
      active_plan: current.active_plan === null
        ? null
        : {
            ...current.active_plan,
            tasks: current.active_plan.tasks.map((item) =>
              item.task_id === task.task_id ? task : item
            ),
          },
    };
  });
}

export function useEditTaskDetails() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      taskId,
      payload,
      idempotencyKey,
    }: {
      taskId: string;
      payload: TaskEditRequest;
      idempotencyKey: string;
    }) => apiRequest<TaskEditResponse>(`/api/v1/tasks/${taskId}/details`, {
      method: "PATCH",
      body: payload,
      idempotencyKey,
    }),
    onSuccess: (result) => invalidateTaskData(qc, result.task.task_id),
  });
}

export function useCreateTaskAdjustmentProposal() {
  return useMutation({
    mutationFn: ({
      taskId,
      version,
      message,
      idempotencyKey,
    }: {
      taskId: string;
      version: number;
      message: string;
      idempotencyKey: string;
    }) => apiRequest<TaskAdjustmentProposalResponse>(
      `/api/v1/tasks/${taskId}/adjustment-proposals`,
      {
        method: "POST",
        body: { version, message },
        idempotencyKey,
      },
    ),
  });
}

export function useConfirmTaskAdjustment() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ adjustmentId, version }: { adjustmentId: string; version: number }) =>
      apiRequest<TaskEditResponse>(
        `/api/v1/task-adjustment-proposals/${adjustmentId}/confirm`,
        { method: "POST", body: { version } },
      ),
    onSuccess: (result) => invalidateTaskData(qc, result.task.task_id),
  });
}

export function useRejectTaskAdjustment() {
  return useMutation({
    mutationFn: ({ adjustmentId, version }: { adjustmentId: string; version: number }) =>
      apiRequest<TaskAdjustmentProposalResponse>(
        `/api/v1/task-adjustment-proposals/${adjustmentId}/reject`,
        { method: "POST", body: { version } },
      ),
  });
}
